pallets/invulnerables/src/lib.rs

```diff
@@ -142,12 +164,39 @@ pub mod pallet {
                 Error::<T>::TooManyInvulnerables
             );
 
-            // should never fail since `new` must be equal to or shorter than `TooManyInvulnerables`
-            let mut bounded_invulnerables = BoundedVec::<_, T::MaxInvulnerables>::try_from(new)
-                .map_err(|_| Error::<T>::TooManyInvulnerables)?;
+            let mut new_with_keys = Vec::new();
+
+            // check if the invulnerables have associated validator keys before they are set
+            for account_id in &new {
+                // don't let one unprepared collator ruin things for everyone.
+                let validator_key = T::CollatorIdOf::convert(account_id.clone());
+                match validator_key {
+                    Some(key) => {
+                        // key is not registered
+                        if !T::CollatorRegistration::is_registered(&key) {
+                            Self::deposit_event(Event::InvalidInvulnerableSkipped {
+                                account_id: account_id.clone(),
+                            });
+                            continue;
+                        }
+                        // else condition passes; key is registered
+                    }
+                    // key does not exist
+                    None => {
+                        Self::deposit_event(Event::InvalidInvulnerableSkipped {
+                            account_id: account_id.clone(),
+                        });
+                        continue;
+                    }
+                }
 
-            // Invulnerables must be sorted for removal.
-            bounded_invulnerables.sort();
+                new_with_keys.push(account_id.clone());
```

Comment:
This loop would be easier to follow without the `continue`s, there are only 3 paths:

```
match key {
    Some(key) if T::CollatorRegistration::is_registered(&key) => {
        // Good, push key
    }
    Some(key) => {
        // Bad, key not registered
    }
    None => {
        // Bad, key does not exist
    }
}
```

Reply:
This was the implementation from the original code. But you are right. I have simplified it in another way.